cmd/disk-cache.go

```diff
@@ -0,0 +1,477 @@
+	return int(usedPercent) > c.maxUsage
+}
+
+// Returns if size space can be allocated without exceeding
+// max disk usage
+func (c diskCache) diskAvailable(size int64) bool {
+	di, err := disk.GetInfo(c.dir)
+	if err != nil {
+		return true
+	}
+	usedPercent := (di.Total - di.Free - uint64(size)) * 100 / di.Total
+	return int(usedPercent) < c.maxUsage
+}
+
+// Purge cache entries that were not accessed.
+func (c diskCache) purge() {
+	delimiter := "/"
+	maxKeys := 1000
+	for {
+		for {
+			if c.diskUsageLow() {
+				break
+			}
+			// expire objects older than one month since accessed
+			expiry := time.Now().AddDate(0, 0, -30)
+			deletedCount := 0
+			buckets, err := c.db.ListBuckets()
+			if err != nil {
+				errorIf(err, "Unable to list buckets.")
+			}
+			for _, bucket := range buckets {
+				var continuationToken string
+				var marker string
+				for {
+					objects, err := c.db.ListObjects(bucket.Name, marker, continuationToken, delimiter, maxKeys)
+					if err != nil {
+						break
+					}
+
+					if !objects.IsTruncated {
+						break
+					}
+					marker = objects.NextMarker
+					for _, object := range objects.Objects {
+						if object.ATime.After(expiry) {
+							continue
+						}
+						if err = c.db.DeleteObject(bucket.Name, object.Name); err != nil {
+							errorIf(err, "Unable to remove cache entry in dir %s/%s", bucket.Name, object.Name)
+						}
+						deletedCount++
+					}
+				}
+			}
+			if deletedCount == 0 {
+				// to avoid a busy loop
+				time.Sleep(time.Second * 30)
+			}
+		}
+		<-c.purgeChan
+	}
+}
+
+// Caches the object in the cache fs layer
+func (c diskCache) Put(bucket, object string, data *hash.Reader, metadata map[string]string, size int64) error {
+	if c.diskUsageHigh() {
+		select {
+		case c.purgeChan <- struct{}{}:
+		default:
+		}
+		return errDiskFull
+	}
+	if !c.diskAvailable(size) {
+		return errDiskFull
+	}
+	if _, err := c.db.GetBucketInfo(bucket); err != nil {
+		pErr := c.db.MakeBucketWithLocation(bucket, "")
+		if pErr != nil {
+			return pErr
+		}
+	}
+	_, err := c.db.PutObject(bucket, object, data, metadata)
+	return err
+}
+
+// Returns the handle for the cached object
+func (c diskCache) Get(bucket, object string, startOffset int64, length int64, writer io.Writer) (err error) {
+	return c.db.GetObject(bucket, object, startOffset, length, writer)
+}
+
+// Returns metadata of the cached object
+func (c diskCache) GetObjectInfo(bucket, object string) (info ObjectInfo, err error) {
+	objInfo, err := c.db.GetObjectInfo(bucket, object)
+
+	if err != nil {
+		return ObjectInfo{}, err
+	}
+	return objInfo, nil
+}
+
+// Deletes the cached object
+func (c diskCache) Delete(bucket, object string) (err error) {
+	return c.db.DeleteObject(bucket, object)
+}
+
+// Inits the cache directory if it is not init'ed already.
+// Initializing implies creation of new FS Object layer and entries directory that holds atime of cached entries
+func newDiskCache(dir string, maxUsage int) (*diskCache, error) {
```

Comment:
Move newThis() to a starting of the function and then implement the necessary calls.